src/Chapter3.hs

```diff
@@ -1006,6 +1119,132 @@ Implement data types and typeclasses, describing such a battle between two
+
+  getAttack :: f a -> Attack
+  receiveAttack :: f a -> Attack -> ActionResult (f a)
+  alive :: f a -> Bool
+
+instance Fighter Monster where
+  actions = monsterActions
+  receiveAttack m (Attack attack) = if alive afterDamage then Alive afterDamage else NotAlive
+      where
+        health = unHealth . monsterHealth $ m
+        afterDamage = m { monsterHealth = Health (health - attack) }
+  getAttack = monsterAttack
+  alive m = (unHealth . monsterHealth $ m) > 0
+
+instance Fighter Knight where
+  actions = knightActions
+  increaseHealth k (IncreaseHealth h) = k { knightHealth = knightHealth k `append` h }
+  increaseDefense k (IncreaseDefense d) = k { knightDefense = knightDefense k `append` d }
+  receiveAttack k (Attack a) = if alive afterDamage then Alive afterDamage else NotAlive
+      where
+        newAttack = a - (unDefense . knightDefense $ k)
+        health = unHealth . knightHealth $ k
+        newHealth = if newAttack < 0 then health else health - newAttack
+        afterDamage = k { knightHealth = Health newHealth }
+  getAttack = knightAttack
+  alive k = (unHealth . knightHealth $ k) > 0
+
+data FightResult a b = FirstWinner a | SecondWinner b deriving (Show, Eq)
+
+data Turn = First | Second
+
+letsFight :: (Fighter f, Action a, Fighter f', Action a') => f a -> f' a' -> FightResult (f a) (f' a')
+letsFight f f' = go (f, cycle (actions f)) (f', cycle (actions f')) First
+  where
+    -- the initial idea was to have one list of all actions taken from first and second fighter one by one
+    -- but I failed to achieve that unless all the actions are of the same type
+    -- then I wanted to just swap second and first on each iteration but still can't achieve that since
+    -- they are of different types though they have the same type class
+    -- and that's why I have this ugly duplication here and the Turn flag which is not very functional I believe
+    go (first, action:restFirstActions) (second, secondActions) First =
+      case target action of
+          Self -> case buff first action of
+                      Alive fir -> go (fir, restFirstActions) (second, secondActions) Second
+                      _ -> SecondWinner second
+          Opponent -> case receiveAttack second (getAttack first) of
+                          Alive sec -> go (first, restFirstActions) (sec, secondActions) Second
+                          _ -> FirstWinner first
+
+    go (first, firstActions) (second, action:restSecondActions) Second =
+      case target action of
+          Self -> case buff second action of
+                      Alive sec -> go (first, firstActions) (sec, restSecondActions) First
+                      _ -> SecondWinner second
+          Opponent -> case receiveAttack first (getAttack second) of
+                          Alive fir -> go (fir, firstActions) (second, restSecondActions) First
+                          _ -> SecondWinner second
+    go l r t = go l r t -- shouldn't be here since actions are infinite
```

Comment:
If you want, you can model your types to make this case impossible. This pattern is called "Making illegal states unrepresentable". You can read more about it here:

* https://kowainik.github.io/posts/haskell-mini-patterns#make-illegal-states-unrepresentable

For example, if you really want to work with only infinite lists, you can create a separate data type for this:

```haskell
data Stream a = Stream a (Stream a)
```

and a helper function:

```haskell
cycleToStream :: [a] -> Stream a  -- though, beware of empty lists
```